From 4eedae0de3f9fe942ebc0eedf037b066789f583e Mon Sep 17 00:00:00 2001
From: Thanh Quan <thanh.quan.xn@renesas.com>
Date: Fri, 30 May 2025 15:40:12 +0700
Subject: [PATCH 44/46] PCIe: dwc: rcar: Integrate MP-PHY into PCIe4 driver

This patch is used to integrate MP-PHY into PCIe4 driver.

Signed-off-by: Thanh Quan <thanh.quan.xn@renesas.com>
---
 drivers/pci/controller/dwc/pcie4-rcar-host.c | 18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

diff --git a/drivers/pci/controller/dwc/pcie4-rcar-host.c b/drivers/pci/controller/dwc/pcie4-rcar-host.c
index 31cbc203513e..9ce0f10273fe 100644
--- a/drivers/pci/controller/dwc/pcie4-rcar-host.c
+++ b/drivers/pci/controller/dwc/pcie4-rcar-host.c
@@ -18,6 +18,7 @@
 #include <linux/reset.h>
 #include <linux/gpio/consumer.h>
 #include <linux/pm_runtime.h>
+#include <linux/phy/phy.h>
 
 #include "pcie-designware.h"
 
@@ -71,6 +72,7 @@
 
 struct rcar_pcie4 {
 	struct dw_pcie			*pci;
+	struct phy			*phy;
 	enum dw_pcie_device_mode	mode;
 	void __iomem			*base;
 	void __iomem			*phy_base;
@@ -94,6 +96,8 @@ static void rcar_gen5_pcie_ltssm_enable(struct rcar_pcie4 *rcar_pcie4,
 	}
 	writel(val, rcar_pcie4->base + PCIERSTCTRL1);
 
+	phy_power_on(rcar_pcie4->phy);
+
 }
 
 static int rcar_gen5_pcie_link_up(struct dw_pcie *pci)
@@ -367,6 +371,13 @@ static int rcar_gen5_pcie_add_port(struct rcar_pcie4 *rcar_pcie4,
 	if (ret) {
 		dev_err(pci->dev, "failed to enable bus clock: %d\n", ret);
 	}
+
+	ret = phy_init(rcar_pcie4->phy);
+	if (ret) {
+		dev_err(dev, "Failed to initialize MP-PHY\n");
+		return ret;
+	}
+
 /*
 	ret = reset_control_deassert(rcar_pcie4->rst);
 	if (ret)
@@ -437,6 +448,13 @@ static int rcar_gen5_pcie_get_resources(struct rcar_pcie4 *rcar_pcie4,
 			rcar_pcie4->phy_base = NULL;
 	}
 
+	/* Get PHY from device tree */
+	rcar_pcie4->phy = devm_of_phy_get_by_index(&pdev->dev, np, 0);
+	if (IS_ERR(rcar_pcie4->phy)) {
+		dev_err(&pdev->dev, "Failed to get PHY: %ld\n", PTR_ERR(rcar_pcie4->phy));
+		return PTR_ERR(rcar_pcie4->phy);
+	}
+
 	return rcar_gen5_pcie_devm_reset_get(rcar_pcie4, &pdev->dev);
 }
 
-- 
2.34.1

